File: client/internal/render/sanitize.go
package render

// Trust-boundary hardening for the terminal surface. The banner content
// (`line`, the OSC-8 click target) is attacker-controlled: it originates from
// the bid API and is rendered into a victim developer's terminal row 1. The web
// tier sanitizes at ingest (apps/web/src/lib/sanitize.ts), but the client is the
// last line of defense — a poisoned creative cache, a compromised server, or a
// MITM must never get a raw ESC/CSI/OSC byte onto the terminal. So we re-strip
// here, mirroring the web ranges, before anything reaches Draw/osc8.

import (
	"net/url"
	"strings"
)

const maxURLLen = 2048

// dangerousFormat reports runes in the bidi-override / zero-width / invisible
// class (the Trojan-Source attack family). Mirrors DANGEROUS_FORMAT in
// apps/web/src/lib/sanitize.ts.
func dangerousFormat(r rune) bool {
	switch {
	case r >= 0x200b && r <= 0x200f, // zero-width + bidi marks
		r >= 0x202a && r <= 0x202e, // bidi embeddings/overrides
		r >= 0x2060 && r <= 0x2064, // word-joiner & invisibles
		r >= 0x2066 && r <= 0x2069, // bidi isolates
		r >= 0x2028 && r <= 0x2029, // line/para separators
		r == 0xfeff:                // BOM / zero-width no-break space
		return true
	}
	return false
}

// controlByte reports runes in the C0 (0x00-0x1F), DEL/C1 (0x7F-0x9F) control
// ranges — every raw ESC/CSI/OSC byte lives here. Mirrors CONTROL in the web tier.
func controlByte(r rune) bool {
	return (r >= 0x00 && r <= 0x1f) || (r >= 0x7f && r <= 0x9f)
}

// sanitizeLine strips control/escape bytes and dangerous-format runes, collapses
// runs of whitespace to a single space, and trims. No raw ESC/CSI/OSC can survive
// so the line cannot break out of row 1 or smuggle a hyperlink/cursor move.
func sanitizeLine(s string) string {
	var b strings.Builder
	b.Grow(len(s))
	prevSpace := false
	for _, r := range s {
		if controlByte(r) || dangerousFormat(r) {
			// Treat a stripped control byte that was whitespace-like (TAB/NL/CR)
			// as a space so words don't fuse; other controls just vanish.
			if r == '\t' || r == '\n' || r == '\r' {
				if !prevSpace {
					b.WriteByte(' ')
					prevSpace = true
				}
			}
			continue
		}
		if r == ' ' {
			if !prevSpace {
				b.WriteByte(' ')
				prevSpace = true
			}
			continue
		}
		b.WriteRune(r)
		prevSpace = false
	}
	return strings.TrimSpace(b.String())
}

// safeURL returns u only if it is a syntactically valid, length-bounded https
// URL whose host contains a dot and which carries no control/escape bytes;
// otherwise "". An empty return makes osc8 degrade to plain (unlinked) text, so
// a hostile click target can never inject an OSC-8 terminator or a non-https
// (file:, javascript:, data:) scheme. Mirrors validateHttpsUrl in the web tier.
func safeURL(u string) string {
	if u == "" || len(u) > maxURLLen {
		return ""
	}
	for _, r := range u {
		if controlByte(r) || dangerousFormat(r) {
			return ""
		}
	}
	parsed, err := url.Parse(u)
	if err != nil {
		return ""
	}
	if parsed.Scheme != "https" {
		return ""
	}
	if !strings.Contains(parsed.Hostname(), ".") {
		return ""
	}
	return u
}

File: apps/web/src/lib/sanitize.ts
/**
 * Creative sanitization. Attacker-controlled bytes (`line`, URLs, `icon`) get
 * rendered into a victim developer's terminal row 1 — a real injection surface.
 * Everything here is defense at the trust boundary (the bid API); the Go client
 * re-clamps width as defense in depth.
 *
 * Hard rules:
 *  - strip ALL control/escape bytes (C0, C1, DEL) so no raw ESC/CSI/OSC survives
 *  - strip bidi overrides + zero-width chars (Trojan-Source class attacks)
 *  - collapse whitespace, bound length
 *  - URLs: https only, parseable, length-bounded; served only via the signed redirect
 *  - icon: small, well-formed image magic bytes, or rejected
 */

export interface CleanCreative {
  line: string;
  displayUrl: string;
  targetUrl: string;
  icon: string | null;
}

export type SanitizeResult =
  | { ok: true; value: CleanCreative }
  | { ok: false; error: string };

const MAX_LINE_CODEPOINTS = 140;
const MAX_URL_LEN = 2048;
const MAX_DISPLAY_LEN = 64;
const MAX_ICON_BYTES = 16 * 1024;

// Build a character-class regex from numeric code-point ranges, so the source
// stays pure ASCII (no literal control/invisible bytes to get mangled).
function charClass(ranges: Array<[number, number]>): RegExp {
  const hex = (n: number) => "\\u" + n.toString(16).padStart(4, "0");
  const body = ranges.map(([a, b]) => (a === b ? hex(a) : `${hex(a)}-${hex(b)}`)).join("");
  return new RegExp(`[${body}]`, "g");
}

// C0 (0x00-0x1F) + DEL (0x7F) + C1 (0x80-0x9F): all control/escape bytes.
const CONTROL = charClass([
  [0x00, 0x1f],
  [0x7f, 0x9f],
]);
// Zero-width, bidi embeddings/overrides/isolates, word-joiner & invisibles,
// line/para separators, BOM — the invisible-character / Trojan-Source attack class.
const DANGEROUS_FORMAT = charClass([
  [0x200b, 0x200f],
  [0x202a, 0x202e],
  [0x2060, 0x2064],
  [0x2066, 0x2069],
  [0x2028, 0x2029],
  [0xfeff, 0xfeff],
]);

/** Strip control + dangerous-format chars, collapse whitespace, trim. */
export function sanitizeText(input: string): string {
  return input
    .replace(CONTROL, "")
    .replace(DANGEROUS_FORMAT, "")
    .replace(/\s+/g, " ")
    .trim();
}

function clampCodepoints(s: string, max: number): string {
  const cps = Array.from(s);
  return cps.length <= max ? s : cps.slice(0, max).join("").trimEnd();
}

/** Validate an https URL; returns the normalized URL or null. */
export function validateHttpsUrl(raw: string): URL | null {
  if (raw.length > MAX_URL_LEN) return null;
  let u: URL;
  try {
    u = new URL(raw);
  } catch {
    return null;
  }
  if (u.protocol !== "https:") return null;
  if (!u.hostname || !u.hostname.includes(".")) return null;
  return u;
}

const IMAGE_MAGIC: Array<{ name: string; bytes: number[] }> = [
  { name: "png", bytes: [0x89, 0x50, 0x4e, 0x47] },
  { name: "jpg", bytes: [0xff, 0xd8, 0xff] },
  { name: "gif", bytes: [0x47, 0x49, 0x46, 0x38] },
];

/** Validate an optional base64 icon: small + recognizable image bytes. */
export function validateIcon(raw: string | null | undefined): { ok: boolean; value: string | null } {
  if (!raw) return { ok: true, value: null };
  const b64 = raw.replace(/^data:image\/[a-z+]+;base64,/, "");
  let buf: Buffer;
  try {
    buf = Buffer.from(b64, "base64");
  } catch {
    return { ok: false, value: null };
  }
  if (buf.length === 0 || buf.length > MAX_ICON_BYTES) return { ok: false, value: null };
  const isRiffWebp =
    buf.length > 12 &&
    buf.toString("ascii", 0, 4) === "RIFF" &&
    buf.toString("ascii", 8, 12) === "WEBP";
  const magicOk =
    isRiffWebp || IMAGE_MAGIC.some((m) => m.bytes.every((b, i) => buf[i] === b));
  if (!magicOk) return { ok: false, value: null };
  return { ok: true, value: b64 };
}

export function sanitizeCreative(input: {
  line: string;
  displayUrl?: string | null;
  targetUrl: string;
  icon?: string | null;
}): SanitizeResult {
  const line = clampCodepoints(sanitizeText(input.line), MAX_LINE_CODEPOINTS);
  if (line.length < 3) return { ok: false, error: "Banner line is empty after sanitizing." };

  const target = validateHttpsUrl(input.targetUrl.trim());
  if (!target) return { ok: false, error: "Target URL must be a valid https:// link." };

  // Prefer a sanitized display URL, else derive a clean host from the target.
  let displayUrl = sanitizeText(input.displayUrl ?? "");
  if (!displayUrl) displayUrl = target.hostname.replace(/^www\./, "");
  displayUrl = clampCodepoints(displayUrl, MAX_DISPLAY_LEN);

  const icon = validateIcon(input.icon);
  if (!icon.ok) return { ok: false, error: "Icon must be a small PNG/JPG/GIF/WebP image." };

  return { ok: true, value: { line, displayUrl, targetUrl: target.toString(), icon: icon.value } };
}
